Add torch.linalg.cholesky_ex without checking for errors by default

IvanYashchuk opened the pull request:
The new function has the following signature `cholesky_ex(Tensor input, *, bool check_errors=False) -> (Tensor L, Tensor infos)`. When `check_errors=True`, an error is thrown if the decomposition fails; `check_errors=False` - responsibility for checking the decomposition is on the user.

When `check_errors=False`, we don't have host-device memory transfers for checking the values of the `info` tensor.

Rewrote the internal code for `torch.linalg.cholesky`. Added `cholesky_stub` dispatch. `linalg_cholesky` is implemented using calls to `linalg_cholesky_ex` now.

Resolves https://github.com/pytorch/pytorch/issues/57032.

Ref. https://github.com/pytorch/pytorch/issues/34272, https://github.com/pytorch/pytorch/issues/47608, https://github.com/pytorch/pytorch/issues/47953

IvanYashchuk committed:
Add torch.linalg.cholesky_ex without checking for errors by default

Rewrote the internal code for `torch.linalg.cholesky`.
Added `cholesky_stub` dispatch. `linalg_cholesky` is implemented to call `linalg_cholesky_ex`.
`torch.linalg.cholesky_ex` suppresses checking of LAPACK error codes by
default and returns them.
That puts the responsibility to check the error codes on the user.

IvanYashchuk committed:
Add tests for check_errors=True

IvanYashchuk committed:
flake8

mruberry commented:
Docs failure is real:

```
Apr 22 21:30:47 Experimental Functions
Apr 22 21:30:47 ---------
Apr 22 21:30:47 /var/lib/jenkins/workspace/docs/source/linalg.rst:47: WARNING: Title underline too short.
```

named tuple list needs an update (so many lists is a pain, sorry):

```
test_native_functions_yaml - TestNamedTupleAPI

Traceback (most recent call last):
  File "test_namedtuple_return_api.py", line 46, in test_native_functions_yaml
    'only allowlisted operators are allowed to have named return type, got ' + name)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1371, in assertEqual
    super().assertTrue(result, msg=self._get_assert_msg(msg, debug_msg=debug_msg))
AssertionError: False is not true : Scalars failed to compare as equal! Comparing 2 and 1 gives a difference of 1, but the allowed difference with rtol=0 and atol=0 is only 0!
only allowlisted operators are allowed to have named return type, got linalg_cholesky_ex
```

mruberry commented:
cc @Balandat, @ngimel, @xwang233

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1337,20 +1334,118 @@ Tensor& cholesky_out(const Tensor &self, bool upper, Tensor &result) {
+    linalg_cholesky_out_info(input, L, info);
+  }
+
+  if (check_errors) {
```
so if someone calls `torch.linalg.cholesky_ex(..., check_errors=True)` and there is an error then the error message would say `torch.linalg.cholesky...`? 

That's not a big deal but it'd probably be easy to pass the calling function's string here?

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1337,20 +1334,118 @@ Tensor& cholesky_out(const Tensor &self, bool upper, Tensor &result) {
+
+std::tuple<Tensor&, Tensor&> linalg_cholesky_ex_out(const Tensor& input, bool check_errors, Tensor& L, Tensor& info) {
+  squareCheckInputs(input);
+  checkSameDevice("torch.linalg.cholesky_ex", L, input, "L");
```
Because cholesky is implemented as a call to cholesky_ex this would show the name cholesky_ex if a call to cholesky hit an error, right? Also seem comment below for when a call to cholesky_ex would show the name cholesky

Error messages aren't the end of the world but do you think we should bother taking a string for the function name? Alternatively we could just always use torch.linalg.cholesky, even when someone calls torch.linalg.cholesky_ex

IvanYashchuk replied:
I agree that we could just use torch.linalg.cholesky.
Here `torch.linalg.cholesky` has its own same checks differing in the function name and that "result" is used instead of "L". It duplicates the check, but that shouldn't have overhead.

IvanYashchuk replied:
So the error messages of `torch.linalg.cholesky` are not changed in this PR. The tests would need to be modified if the error messages would change.

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -8609,17 +8609,26 @@
+    CompositeExplicitAutograd: linalg_cholesky_ex
 
-- func: linalg_cholesky.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
+- func: linalg_cholesky_ex.L(Tensor self, *, bool check_errors=False, Tensor(a!) L, Tensor(b!) info) -> (Tensor(a!) L, Tensor(b!) info)
```
.out?

mruberry replied:
Jumping ahead of @ezyang here: could we look at making these structured? If that seems challenging we could look at doing it in a follow-up PR.

ezyang replied:
I don't see a CPU, CUDA dispatch key here so this might be an alias style kernel that may not work with structured. You should add this to the binder of cases to solve with the transpiler.

IvanYashchuk replied:
Is `CPU, CUDA` dispatch key preferred over `CompositeExplicitAutograd` if the dispatch stubs are used?

I haven't looked at porting to structured in detail, so I can't estimate how challenging is it. I was planning to port linalg module functions to structured after the 1.9 branch cut.

mruberry commented on test/test_linalg.py:
```diff
@@ -575,6 +575,63 @@ def run_test(upper):
+    @skipCUDAIfNoMagmaAndNoCusolver
+    @skipCPUIfNoLapack
+    @dtypes(torch.float32, torch.float64, torch.complex64, torch.complex128)
+    def test_cholesky_with_info(self, device, dtype):
```
test_cholesky_ex? That will make it easier for people to find the test(s) associated with the operator

mruberry commented on test/test_linalg.py:
```diff
@@ -575,6 +575,63 @@ def run_test(upper):
+        from torch.testing._internal.common_utils import random_hermitian_pd_matrix
+
+        def run_test(n, batch):
+            A = random_hermitian_pd_matrix(n, *batch, dtype=dtype, device=device)
```
I'm curious about these tests because we already have a lot of tests for cholesky, and cholesky_ex and cholesky are "near aliases" of each other; should we just refactor the cholesky tests so that test_cholesky_foo and test_cholestky_foo_ex both call a common helper, but pass a different torch function? Then there can be separate tests that validate their divergent behavior on infos

IvanYashchuk replied:
I completely agree. I think it would take a lot of energy from me to refactor the tests now. If those functions were aliases the task of course would be easy. I'd prefer to do it in a follow-up PR, after 1.9 branch cut, if that's okay.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -84,6 +84,54 @@
     True
 """)
 
+cholesky_ex = _add_docstr(_linalg.linalg_cholesky_ex, r"""
```
I like that we're documenting this completely separately from linalg.cholesky because it's an experimental function

mruberry replied:
That said, we should probably do two things:

- notify the user this function is a PROTOTYPE that may change in a future PyTorch release
- explain how it's different from torch.linalg.cholesky (it always returns the infos, with check_errors=False the computation is the same, with check_errors=True the computation skips error checking, which can be useful if users would prefer to do the error checking themselves or they know the inputs will not throw an error)

IvanYashchuk replied:
I added a warning, is it okay?
```
.. warning:: This function is "experimental" and it may change in a future PyTorch release.
```
Not sure on the use of quotes though.

mruberry commented on test/test_linalg.py:
```diff
@@ -575,6 +575,63 @@ def run_test(upper):
+    @skipCUDAIfNoMagmaAndNoCusolver
+    @skipCPUIfNoLapack
+    @dtypes(torch.float32, torch.float64, torch.complex64, torch.complex128)
+    def test_cholesky_with_info_non_pd(self, device, dtype):
```
This test is very cool

mruberry commented on docs/source/linalg.rst:
```diff
@@ -42,3 +42,11 @@ Functions
     lstsq
     householder_product
+
+Experimental Functions
```
Good idea

mruberry commented:
Hey @IvanYashchuk! This is awesome!

I made a few comments about error messages, a docs tweak, a possible test refactor, and using structure kernels. I don't think any of them are blocking, we just need to focus on fixing the tests and any additional tweaks are a bonus. We can pursue the docs as part of our docs review, for example.

@Lezcano would you like to sanity check the impl, too? 

Let's also give @ezyang an opportunity to look at the native_functions.yaml changes

Balandat commented:
This is great @IvanYashchuk. Let me see if using this and checking the `info` resolves glacial speed issues reported in https://github.com/pytorch/pytorch/issues/34272#issue-575895944?

ngimel commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1659,65 +1672,89 @@ AT_ERROR("cholesky: MAGMA library not found in "
     // which concisely is equal to batch_size % batch_limit
     if (batch_size % batch_limit != 0) {
       magmaCholeskyBatched<scalar_t>(
```
nit, but consider bringing this to the upper loop, the only change needed is `mini_idx < batch_size` in the loop statement, and `nbatches = std::min(batch_limit, batch_size - mini_idx)` as argument to magmaCholeskyBatched.

ngimel commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1659,65 +1672,89 @@ AT_ERROR("cholesky: MAGMA library not found in "
 Tensor _cholesky_helper_cuda(const Tensor& self, bool upper) {
+  auto info_shape = IntArrayRef(
+    self.sizes().cbegin(), self.sizes().cend() - 2); // self.shape[:-2]
+  Tensor self_working_copy = cloneBatchedColumnMajor(self);
```
can you please leave a comment here that clone is called because legacy cholesky doesn't do clone before?

Balandat commented:
This is great. In this notebook I made a quick and dirty update of the `psd_safe_cholesky` helper we're using in gpytorch that automatically adds jitter upon failure of cholesky: [test_cholesky_ex.ipynb.zip](https://github.com/pytorch/pytorch/files/6362951/test_cholesky_ex.ipynb.zip)

For a small matrix example this gets the runtime down from 123ms -> 76µs. Yes, those are the correct units, this is a **speedup of ~2000X**. 

cc @jacobrgardner, @gpleiss, @dme65, @vishwakftw

xwang233 commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1659,65 +1672,89 @@ AT_ERROR("cholesky: MAGMA library not found in "
+  if (batchCount(input) == 1 || !use_magma_) {
+    cholesky_helper_cusolver(input, upper, info);
+  } else {
+    cholesky_helper_magma(input, upper, info);
```
Do we know if `cholesky_helper_magma` syncs with CPU internally?

IvanYashchuk replied:
Yes, we know that it does for single matrix input, and for batched cases, it has cudaMalloc calls. https://github.com/pytorch/pytorch/issues/47953#issuecomment-762389928

xwang233 commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1337,20 +1334,118 @@ Tensor& cholesky_out(const Tensor &self, bool upper, Tensor &result) {
+  copy_needed |= (info.numel() != 0 && !(info.sizes().equals(expected_info_shape))); // or L does not have the expected shape
+  copy_needed |= (info.scalar_type() != info_output_type);  // or L does not have the same dtype as input
+
+  if (copy_needed) {
```
I see that input, L, info, L_tmp, info_tmp are all on the same device. Is there any chance that a resize or a copy_ could cause CUDACachingAllocator to malloc (or cudaMalloc) internally and trigger a CPU sync?

IvanYashchuk replied:
I don't know the internals of CUDACachingAllocator.

ngimel replied:
CUDACachingAllocator can always issue cudaMalloc, for any operation, we don't note that as a potential synchronization source.

xwang233 commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1659,65 +1672,89 @@ AT_ERROR("cholesky: MAGMA library not found in "
+  }
+#else
+  cholesky_helper_magma(input, upper, info);
+#endif // USE_CUSOLVER
```
If `cholesky_kernel` and `_cholesky_helper_cuda` will be using the same heuristic, it would be really nice to put them into once place.

IvanYashchuk replied:
Yes, but `cholesky_kernel` in its current form doesn't accept `bool upper`. It's easy to change, but it was done on purpose since `torch.linalg.cholesky` doesn't support it and I wanted to keep `torch.cholesky` and `torch.linalg.cholesky` separate, to introduce minimal changes.

Looking at it now I can probably add `bool upper` to `cholesky_kernel` and implement all the cholesky functions using the same `cholesky_stub`. But I should probably split this PR.

xwang233 reviewed (commented):
Thanks for the great work @IvanYashchuk ! I've left some minor comments for the code. 

One concern I have is that do we know if `magma_cholesky` syncs with CPU? If it syncs with CPU, should we use cusolver dispatch only for `check_errors = False` ?

IvanYashchuk commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1659,65 +1672,89 @@ AT_ERROR("cholesky: MAGMA library not found in "
+// The same MAGMA/cuSOLVER dispatch logic is used here as in _cholesky_helper_cuda
+// This function is for cholesky_stub that is used for linalg_cholesky_ex
+// while _cholesky_helper_cuda is used for legacy torch.cholesky
+void cholesky_kernel(const Tensor& input, const Tensor& info) {
```
Yes, it's correct. It's an in-place function taking const references to tensors whose storage is modified. See https://github.com/pytorch/rfcs/blob/rfc-0005/RFC-0005-structured-kernel-definitions.md#removing-mutable-references-from-out-arguments

IvanYashchuk commented:
> One concern I have is that do we know if magma_cholesky syncs with CPU? If it syncs with CPU, should we use cusolver dispatch only for check_errors = False ?

The main goal for this function is not to avoid the synchronization, but give users the possibility to suppress the error check and return useful information about the input matrices contained in lapack error codes.
We should use the cusolver dispatch independent of check_errors because it's faster for inputs with single matrix.

IvanYashchuk commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1337,20 +1334,118 @@ Tensor& cholesky_out(const Tensor &self, bool upper, Tensor &result) {
+  checkLinalgCompatibleDtype("torch.linalg.cholesky_ex", L, input, "L");
+  checkSameDevice("torch.linalg.cholesky_ex", info, input, "info");
+  ScalarType info_output_type = ScalarType::Int;
+  checkLinalgCompatibleDtype("torch.linalg.cholesky_ex", info.scalar_type(), info_output_type);
```
I think `info` should follow the same behavior that is expected for all other "out" tensors. It should be possible to provide a tensor with different castable dtype, tensors with zero elements are resized, etc. Is that right?

ngimel replied:
Not exactly. We are limiting type promotion behavior to pointwise functions and reductions #56356. Existing additional functionality is grandfathered in, but we should avoid adding new. So, linalg functions casting to output exists today, but `info` is a new addition, and should not be promoted.

xwang233 commented:
> The main goal for this function is not to avoid the synchronization, but give users the possibility to suppress the error check and return useful information about the input matrices contained in lapack error codes.
We should use the cusolver dispatch independent of check_errors because it's faster for inputs with single matrix.

Ok, I see, that makes sense. We can add cusolver potrf batched in the future once the issue in https://github.com/pytorch/pytorch/issues/53879 gets fixed. With that, we should be able to use cusolver potrf + potrfBatched to fully remove the synchronization behavior in Cholesky.

xwang233 commented on torch/linalg/__init__.py:
```diff
@@ -84,6 +84,54 @@
+``info`` filled with zeros indicates that the decomposition was successful.
+If ``check_errors=True`` and ``info`` contains positive integers, then a RuntimeError is thrown.
+
+.. note:: If ``check_errors=True`` given inputs on a CUDA device, this function synchronizes that device with the CPU.
```
Users may concern what could happen with `check_errors=False`. Do we also want to mention that behavior? Like, *the function may synchronize that device with the CPU*, without giving too many details?

vishwakftw commented:
This is wonderful stuff, thanks for implementing this @IvanYashchuk !

IvanYashchuk commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -8609,17 +8609,26 @@
   variants: function
   dispatch:
-    CompositeExplicitAutograd: linalg_cholesky
+    CompositeExplicitAutograd: linalg_cholesky_ex
```
Change this to `CPU, CUDA` dispatch key.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into cholesky-check-errors

vadimkantorov commented:
Out of curiosity, what's the motivation behind a new `cholesky_ex` versus just adding an argument `check_errors`? If this pattern gets expanded, we'll end up with a lot of `_ex` functions :( and that would be confusing. 

If this is to allow incomplete Cholesky, should we just call the function `cholesky_incomplete` and not return the `infos`? Are `infos` useful themselves, besides just suppressing the checking?

ngimel commented:
Motivation is #46187, which is apparently unsolvable in other ways, so we cannot return different number of outputs depending on the values of other kwargs.

mruberry commented:
> Out of curiosity, what's the motivation behind a new `cholesky_ex` versus just adding an argument `check_errors`? If this pattern gets expanded, we'll end up with a lot of `_ex` functions :( and that would be confusing.

We didn't want to condition the # of tensors returned on the input arguments.

vadimkantorov commented:
In general I understand. But in this concrete case, is it needed only for https://github.com/pytorch/pytorch/issues/56921, then keeping the single-tensor-returning version may be okay

IvanYashchuk commented:
> Are infos useful themselves, besides just suppressing the checking?

`info` tensor is useful because it tells the user which matrices are not positive definite. Running Cholesky decomposition is one of the fastest ways to obtain such information algorithmically, but it wasn't fast practically (https://github.com/pytorch/pytorch/issues/34272).

That's an unfortunate situation that we end up with three functions computing the Cholesky decomposition for dense matrices: `torch.cholesky`, `torch.linalg.cholesky`, and `torch.linalg.cholesky_ex`. However, `torch.cholesky` is (or will be in 1.9 release) deprecated in favor of `torch.linalg.cholesky` and will be removed in the future.
`torch.linalg.cholesky` is a NumPy compatible function and NumPy compatibility was one of the main motivations for `torch.linalg` namespace.
Working on `torch.linalg` module we noticed a few inconsistencies and things that could be improved. `torch.linalg.cholesky_ex{perimental}` represents one idea that we think is a useful addition: some LAPACK-based function return additional metadata about the inputs and this metadata is useful and shouldn't be discarded. The new function signature for cholesky will be proposed for [the Array API standard](https://github.com/data-apis/array-api/) and, if accepted, will be implemented in NumPy in the future.

vadimkantorov commented:
One way may be to allow the user to pass in extra arguments by reference e.g. `infos = torch.Tensor(); toch.linalg.cholesky(..., infos = infos); print(infos)`, like we do with out-args.

Balandat commented:
> One way may be to allow the user to pass in extra arguments by reference e.g. `infos = torch.Tensor(); toch.linalg.cholesky(..., infos = infos); print(infos)`, like we do with out-args.

I like that idea. Maybe just have `info` be a dict that gets a `lapack_info` key set when passed in?

IvanYashchuk commented:
> > One way may be to allow the user to pass in extra arguments by reference e.g. `infos = torch.Tensor(); toch.linalg.cholesky(..., infos = infos); print(infos)`, like we do with out-args.
> 
> I like that idea. Maybe just have `info` be a dict that gets a `lapack_info` key set when passed in?

There was a draft PR implementing this for `torch.linalg.inv` https://github.com/pytorch/pytorch/pull/52046. There are some technical difficulties with that approach and I think for many Python users it is not intuitive to pass tensors by reference to be filled. It can be revisited in the future and we could implement such thing for `torch.linalg.cholesky`, it is calling `cholesky_ex` internally anyway.

Balandat commented:
> There was a draft PR implementing this for `torch.linalg.inv` #52046. There are some technical difficulties with that approach and I think for many Python users it is not intuitive to pass tensors by reference to be filled. It can be revisited in the future and we could implement such thing for `torch.linalg.cholesky`, it is calling `cholesky_ex` internally anyway.

So I was thinking something very simple that doesn't require passing a tensor by reference, sth like the following:
```
from typing import Dict, Optional
from torch import Tensor

import torch


def cholesky(M: Tensor, diagnostics: Optional[Dict[str, Tensor]] = None) -> Tensor:
    L, info = torch.linalg.cholesky_ex(M)
    if diagnostics is not None:
        diagnostics["info"] = info
    return L
```
I agree this is a bit of a weird interface, but it would help reduce proliferation of functions, and since anyone using this kind of functionality should know what they're doing imo the complexity from a user perspective shouldn't be disqualifying for this.

vadimkantorov commented:
At least for the `infos` case, this could solve the problem of having several function versions. The general API case still rests unsolved, but maybe that could be dealt with when another realistic usecase appears.

IvanYashchuk committed:
Replace CompositeExplicitAutograd with CPU, CUDA

dispatch stubs are used and they are implementedo only for CPU and CUDA

IvanYashchuk committed:
Replace TORCH_INTERNAL_ASSERT -> TORCH_INTERNAL_ASSERT_DEBUG_ONLY

IvanYashchuk committed:
Restrict dtype of info tensor

IvanYashchuk committed:
Add a comment about the clone for legacy cholesky

IvanYashchuk committed:
Simplify looping over batched input beyond MAGMA's batch limit

IvanYashchuk committed:
Fix batched magma with upper=true

IvanYashchuk committed:
Make the synchronization note more general

IvanYashchuk commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1619,17 +1621,28 @@ AT_ERROR("cholesky: MAGMA library not found in "
+    // if upper=true we need to tranpose the self tensor
+    if (upper) {
+      // self.transpose_(-2, -1);
+      self_data = self.transpose(-2, -1).data_ptr<scalar_t>();
```
Looking at it now, this portion of the code is wrong and I am removing it. Thank you!
The original intention was trying to solve https://github.com/pytorch/pytorch/issues/57032. This was wrong and the issue is resolved now with changes to the `cholesky_helper_magma` function.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into cholesky-check-errors

IvanYashchuk committed:
Fix Title underline too short

IvanYashchuk committed:
Add namedtuple test

IvanYashchuk committed:
Rename test_cholesky_with_info -> test_cholesky_ex

IvanYashchuk committed:
Add a warning that cholesky_ex is experimental

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into cholesky-check-errors

IvanYashchuk committed:
Fix complex input workaround for upper=true

IvanYashchuk committed:
Remove _cholesky_helper

Add upper argument to cholesky_stub

IvanYashchuk committed:
Add test for issue https://github.com/pytorch/pytorch/issues/57032

IvanYashchuk commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1272,52 +1269,28 @@ Tensor& cholesky_solve_out(const Tensor& self, const Tensor& A, bool upper, Tens
 // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ cholesky ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-template<typename scalar_t>
-static void apply_cholesky(Tensor& self, bool upper, std::vector<int64_t>& infos) {
```
This function is moved to `BatchLinearAlgebraKernel.cpp` file.

IvanYashchuk commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1272,52 +1269,28 @@ Tensor& cholesky_solve_out(const Tensor& self, const Tensor& A, bool upper, Tens
   }
   squareCheckInputs(self);
 
-  auto raw_cholesky_output = at::_cholesky_helper(self, upper);
```
Removed `at::_cholesky_helper`, it is replaced with `cholesky_stub`.

IvanYashchuk committed:
Fix function names in the error messages

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into cholesky-check-errors

IvanYashchuk commented:
@mruberry, I updated this pull request. Most of the changes in this PR are rewrites of internal code for cholesky needed to expose the `info` tensor and using dispatch stubs instead of aten::helper.

`torch.cholesky`, `torch.linalg.cholesky`, `torch.linalg.cholesky_ex` have their own error checks so that the user-facing function name is used.
I removed `at::_cholesky_helper`, it is replaced with `cholesky_stub`.
Fixed the bug in `torch.cholesky` (https://github.com/pytorch/pytorch/issues/57032) and added a test for it.

mruberry reviewed (approved):
@ngimel and I had a chance to take a look, and this is awesome!

It just needs to be rebased.

@xwang233, what's the status on cusolverDnXpotrfBatched, by the way?

xwang233 commented:
@mruberry For `cusolverDnXpotrfBatched`, it has been fixed internally and will be released soon in the upcoming cuda releases! I'll keep you guys updated when it's ready. Thanks.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into cholesky-check-errors

IvanYashchuk commented:
@mruberry, thanks!
I resolved merge conflicts.
